### PR TITLE
many: finalize trusty support

### PR DESCRIPTION
**Pull request by vosst**
dh-systemd is not usable on trusty due to the deputy systemd instance considering different unit directories. For that, we resort to manually enabling/starting/stopping units.

We introduce a unit snap.mount.service that takes care of selectively enabling rshared support for /snap on trusty.

Finally, a couple of remaining changes to the spread test suite are required

### CODE REVIEW

File: debian/rules [hunk 1]
```diff
@@ -145,10 +103,14 @@ override_dh_install:
 	install --mode=0644 debian/*.socket debian/snapd/$(SYSTEMD_UNITS_DESTDIR)
 	install --mode=0644 debian/snapd.service debian/snapd/$(SYSTEMD_UNITS_DESTDIR)
 	install --mode=0644 debian/*.target debian/snapd/$(SYSTEMD_UNITS_DESTDIR)
-ifeq ($(RELEASE),trusty)
-	dh_link debian/snapd/$(SYSTEMD_UNITS_DESTDIR)/snapd.service debian/snapd/$(SYSTEMD_UNITS_DESTDIR)/multi-user.target.wants/snapd.service
-	dh_link debian/snapd/$(SYSTEMD_UNITS_DESTDIR)/snapd.autoimport.service debian/snapd/$(SYSTEMD_UNITS_DESTDIR)/multi-user.target.wants/snapd.autoimport.service
-endif
+	install --mode=0644 debian/snapd.boot-ok.service debian/snapd/$(SYSTEMD_UNITS_DESTDIR)
+	install --mode=0644 debian/snapd.firstboot.service debian/snapd/$(SYSTEMD_UNITS_DESTDIR)
+	dh_link /$(SYSTEMD_UNITS_DESTDIR)/snapd.firstboot.service /$(SYSTEMD_UNITS_DESTDIR)/multi-user.target.wants/snapd.firstboot.service
```
Comment (chipaca) [inline debian/rules]: instead of manually shipping the symlinks I'd much rather you used `systemctl enable`. You can even `systemctl enable --now $UNIT` to start it at the same time as you enable it.

File: debian/snapd.postinst [hunk 2]
```diff
@@ -11,4 +11,13 @@ case "$1" in
...
             ldconfig
         fi
+        # start required services
+        if [ -d /run/systemd/system ]; then
+            systemctl start snapd.boot-ok.service
+            systemctl start snapd.refresh.timer
+            systemctl start snapd.refresh.service
+            systemctl start snapd.autoimport.service
+            systemctl start snapd.socket
+            systemctl start snapd.service
```
Comment (chipaca) [inline debian/snapd.postinst]: I think you need to `systemctl daemon-reload` in there, also.

File: debian/snapd.postinst [hunk 3]
```diff
@@ -11,4 +11,13 @@ case "$1" in
         if dpkg --compare-versions "$2" lt-nl "2.0.7"; then
             ldconfig
         fi
+        # start required services
+        if [ -d /run/systemd/system ]; then
+            systemctl start snapd.boot-ok.service
+            systemctl start snapd.refresh.timer
+            systemctl start snapd.refresh.service
```
Comment (chipaca) [inline debian/snapd.postinst]: no, don't start the refresh service manually; start the timer.

File: debian/snapd.prerm [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#! /bin/sh
+# prerm script for snapd
+
+set -e
+
+if [ "$1" = "remove" -o "$1" = "upgrade" ]; then
```
Comment (chipaca) [inline debian/snapd.prerm]: why do you not want to try to stop services unless removing or upgrading? the other ways the prerm script is called is when you most need help. Wrap each one in `systemctl is-active --quiet $UNIT`.

File: debian/snapd.postinst (hunk 2 above)
Comment (chipaca) [inline debian/snapd.postinst]: Also, given you're doing it manually, could you please do a single call to systemctl instead of N? It'll happily take all the unit files you want to perform the same action in a single call

File: debian/snapd.prerm (hunk 4 above)
Comment (chipaca) [inline debian/snapd.prerm]: Also, consider whether you mean just stop, or stop-and-disable, here (the latter is `disable --now`)

File: debian/rules (hunk 1 above)
Comment (vosst) [inline debian/rules]: Fixed, --now is not available in trusty's systemctl, though.

File: debian/snapd.postinst (hunk 3 above)
Comment (vosst) [inline debian/snapd.postinst]: Fixed.

File: debian/snapd.postinst (hunk 2 above)
Comment (vosst) [inline debian/snapd.postinst]: Both fixed.

File: debian/snapd.prerm (hunk 4 above)
Comment (vosst) [inline debian/snapd.prerm]: Fixed according to your first comment, and resorted to stop and disable. --now is not available as pointed out in an earlier comment.

File: debian/rules [hunk 5]
```diff
@@ -9,6 +9,8 @@ export DH_GOLANG_EXCLUDES=tests
 export DH_GOLANG_GO_GENERATE=1
 
 export PATH:=${PATH}:${CURDIR}
+# make sure that correct go version is found on trusty
+export PATH:=/usr/lib/go-1.6/bin:${PATH}
 
 RELEASE = $(shell lsb_release -c -s)
```
Comment (zyga) [inline debian/rules]: Can we please use `/etc/os-release` instead?

File: debian/control [hunk 6]
```diff
@@ -10,8 +10,8 @@ Build-Depends: bash-completion,
                gettext,
                gnupg2,
                golang-any (>=2:1.6) | golang-1.6,
-               golang-github-gosexy-gettext-dev,
                python3,
+               lsb-release,
```
Comment (zyga) [inline debian/control]: Can we please drop this and use `/etc/os-release` instead?

**Comment by chipaca** [comment]
Give me a shout on telegram when this is ready for me to take a look at it again please

File: spread.yaml [hunk 7]
```diff
@@ -103,13 +100,15 @@ debug-each: |
...
     # this indicates that the server got reused, nothing to setup
     [ "$REUSE_PROJECT" != 1 ] || exit 0
-
     # check that we are not updating
     . "$TESTSLIB/boot.sh"
     if [ "$(bootenv snap_mode)" = "try" ]; then
         echo "Ongoing reboot upgrade process, please try again when finished"
         exit 1
     fi
+    # XXX: hack
```
Comment (zyga) [inline spread.yaml]: This should be dropped. If tests fail without this they will not represent what actual users will see.

Comment (vosst) [inline spread.yaml]: Thanks for the reminder, hack dropped, tests passing locally. Let's see what the machinery thinks.

File: tests/main/interfaces-fuse_support/task.yaml [hunk 8]
```diff
@@ -1,5 +1,7 @@
 summary: Ensure that the fuse-support interface works.
 
+systems: [-ubuntu-14.04-64]
+
```
Comment (vosst) [inline tests/main/interfaces-fuse_support/task.yaml]: What would be the best forward? Only conditionally include it in interfaces/builtin/all.go?

**Comment by martinpitt** [comment]
Thomas asked me to trigger an extra trusty/amd64 run.

File: #spread.yaml# [hunk 9]
```diff
@@ -0,0 +1,229 @@
+project: snapd
```
Comment (zyga) [inline #spread.yaml#]: Did you really intend to add `#spread.yaml#`?

**Comment by martinpitt** [comment]
trusty failed as the snappy-dev/image PPA does not have any indexes for trusty. @mvo5, can you please do the usual "binary copy/publish/remove" dance?

**Comment by zyga** [comment]
Can you please merge master into your branch?

File: debian/control [hunk 10]
```diff
@@ -10,8 +10,8 @@ Build-Depends: bash-completion,
                gettext,
                gnupg2,
                golang-any (>=2:1.6) | golang-1.6,
-               golang-github-gosexy-gettext-dev,
```
Comment (vosst) [inline debian/control]: Not any more with vendoring in place.

File: debian/snap.mount.service [hunk 11]
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description=Apply special mount sharing to /snap directory
+ConditionPathExists=!/run/systemd/initctl/fifo
```
Comment (vosst) [inline debian/snap.mount.service]: /run/systemd/initctl/fifo only exists if systemd runs as PID 1.

File: spread.yaml [hunk 12]
```diff
@@ -92,9 +90,9 @@ exclude:
     - .git
 
 prepare-each: |
-    journalctl --rotate
+    journalctl --rotate || true
```
Comment (vosst) [inline spread.yaml]: --rotate and --vaccum-time are not available on trusty's systemd.

File: debian/rules [hunk 13]
```diff
@@ -9,11 +9,13 @@ export DH_GOLANG_EXCLUDES=tests
 export DH_GOLANG_GO_GENERATE=1
 
 export PATH:=${PATH}:${CURDIR}
+# make sure that correct go version is found on trusty
+export PATH:=/usr/lib/go-1.6/bin:${PATH}
 
-RELEASE = $(shell lsb_release -c -s)
+include /etc/os-release
```
Comment (zyga) [inline debian/rules]: ❤️

File: debian/rules [hunk 14]
```diff
@@ -27,6 +29,8 @@ else
 	SYSTEMD_UNITS_DESTDIR="lib/systemd/system/"
 endif
 
+$(info ${SYSTEMD_UNITS_DESTDIR})
```
Comment (zyga) [inline debian/rules]: This looks like leftover debugging, can we please remove it?

File: debian/snapd.postrm [hunk 15]
```diff
@@ -13,6 +13,19 @@ systemctl_stop() {
 if [ "$1" = "purge" ]; then
     mounts=$(systemctl list-unit-files | grep '^snap[-.].*\.mount' | cut -f1 -d ' ')
     services=$(systemctl list-unit-files | grep '^snap[-.].*\.service' | cut -f1 -d ' ')
+
+    for unit in $mounts; do
+        # ensure its really a snapp mount unit or systemd unit
```
Comment (zyga) [inline debian/snapd.postrm]: Typo `snapp` should be `snap`

File: debian/snapd.postrm [hunk 16]
```diff
@@ -41,9 +54,9 @@ if [ "$1" = "purge" ]; then
             rm -rf "/var/snap/$snap/common"
             rm -f "/var/snap/$snap/current"
             # opportunistic remove (may fail if there are still revisions left
-            for d in "/snap/bin" "/snap/$snap" "/var/snap/$snap" "/snap" "/var/snap"; do
+            for d in "/snap/bin" "/snap/$snap" "/var/snap/$snap" "/var/snap"; do
                 if [ -d "$d" ]; then
-                    rmdir --ignore-fail-on-non-empty $d
+                    rmdir --ignore-fail-on-non-empty $d || true
```
Comment (zyga) [inline debian/snapd.postrm]: Since we removed `/snap` from the loop why do we need the `|| true` part?

File: spread.yaml (hunk 12 above)
Comment (zyga) [inline spread.yaml]: Can you please add that as a comment instead

File: debian/control [hunk 17]
```diff
@@ -40,10 +40,13 @@ Package: snapd
 Architecture: any
 Depends: adduser,
          ca-certificates,
+         cgroup-lite,
          gnupg1 | gnupg,
          snap-confine (>= 1.0.43),
          squashfs-tools,
+         sudo,
```
Comment (vosst) [inline debian/control]: I'm using the snapd package to force installation/update of certain packages. In the case of sudo, this (more precisely, a versioned dependency on sudo) forces installation of an update that adds /snap/bin to the secure path.

File: debian/snapd.postrm (hunk 16 above)
Comment (vosst) [inline debian/snapd.postrm]: We don't need and we want to be strict here. Will remove || true.

File: debian/snapd.postrm [hunk 18]
```diff
@@ -33,17 +43,17 @@ if [ "$1" = "purge" ]; then
             rm -f "/snap/bin/$snap"
             rm -f "/snap/bin/$snap".*
             # snap mount dir
-            umount -l "/snap/$snap/$rev" 2> /dev/null || true
+            umount -d -l "/snap/$snap/$rev" 2> /dev/null || true
```
Comment (zyga) [inline debian/snapd.postrm]: Interesting, I  didn't know about `-d`. Is this something we should be doing ourselves elsewhere?

File: debian/snapd.postrm (hunk 16 above)
Comment (zyga) [inline debian/snapd.postrm]: How can it be non-empty if we remove components of `/var/snap/` above? Sorry for nitpicking but I'd rather know what happens here.

File: spread.yaml [hunk 19]
```diff
@@ -32,15 +34,13 @@ backends:
             - ubuntu-core-16-64:
                 kernel: Direct Disk
                 image: ubuntu-16.04-64
-            # FIXME restore of ubuntu-core does not properly reset
```
Comment (zyga) [inline spread.yaml]: Why is this being removed?

File: spread.yaml [hunk 20]
```diff
@@ -75,7 +74,6 @@ backends:
         type: adhoc
         environment:
             SPREAD_EXTERNAL_ADDRESS: "$(HOST: echo ${SPREAD_EXTERNAL_ADDRESS:-localhost:8022})"
-            MANAGED_DEVICE: "true"
```
Comment (zyga) [inline spread.yaml]: Why is this being removed?

File: tests/main/gccgo/task.yaml [hunk 21]
```diff
@@ -1,5 +1,5 @@
 summary: Check that snapd builds with gccgo
-systems: [-ubuntu-core-16-64, -ubuntu-core-16-arm-64, -ubuntu-core-16-arm-32]
+systems: [-ubuntu-core-16-64, -ubuntu-core-16-arm-64, -ubuntu-core-16-arm-32, -ubuntu-14.04-64]
```
Comment (zyga) [inline tests/main/gccgo/task.yaml]: Did we adjust the packaging to not build on ppc64el if there's no gccgo available?

File: debian/snapd.postrm (hunk 18 above)
Comment (vosst) [inline debian/snapd.postrm]: I have to verify this again, but I think that util-linux is doing the right thing on newer versions.

File: debian/snapd.postrm (hunk 16 above)
Comment (vosst) [inline debian/snapd.postrm]: We remove per iteration of the loop, so this tries to remove per iteration and hopefully succeeds at some point.

File: spread.yaml (hunk 19 above)
Comment (vosst) [inline spread.yaml]: Not on purpose, likely happened during a merge. restored.

File: spread.yaml [hunk 22]
```diff
@@ -97,9 +102,12 @@ exclude:
     - .git
 
 prepare-each: |
-    journalctl --rotate
+    # systemd on 14.04 does not know about --rotate
+    # or --vacuum-time. Gracefully handle both commands
+    # failing.
```
Comment (niemeyer) [inline spread.yaml]: We don't want to ignore errors on those. We can do this instead:
```
if [ "$SPREAD_SYSTEM" = ubuntu-14.04-64 ]; then
...
else
...
fi
```

Then, we need to find something to put on that if clause for 14.04 rather than just disabling it, because otherwise the content of `debug-each` below will cause a ridiculous amount of unrelated logs to be shown on errors, to the point of making spread non-useful.

There must be a way to clean logs on 14.04, right?

File: snap/implicit.go [hunk 23]
```diff
@@ -27,41 +27,50 @@ import (
...
-	"removable-media",
-	"shutdown",
-	"snapd-control",
-	"system-observe",
-	"system-trace",
-	"time-control",
-	"timeserver-control",
-	"timezone-control",
-	"tpm",
+func initImplicitSlots() []string {
```
Comment (niemeyer) [inline snap/implicit.go]: No need to change here. See below.. we're already building up the final list inside `AddImplicitSlots`. Let's just add the `fuse-support` there conditionally.

File: interfaces/builtin/all.go [hunk 24]
```diff
@@ -21,72 +21,82 @@ package builtin
...
-	NewSystemObserveInterface(),
-	NewSystemTraceInterface(),
-	NewTimeControlInterface(),
-	NewTimeserverControlInterface(),
-	NewTimezoneControlInterface(),
-	NewTpmInterface(),
-	NewUPowerObserveInterface(),
-	NewUnity7Interface(),
-	NewX11Interface(),
+func initInterfaces() []interfaces.Interface {
```
Comment (niemeyer) [inline interfaces/builtin/all.go]: Can we avoid changing this?  Ideally `allInterfaces` would remain being _all interfaces._ so that snapd has metadata about all the interfaces it supports, even if it's unsupported in the local system.

File: debian/snapd.postrm [hunk 25]
```diff
@@ -13,15 +13,25 @@ systemctl_stop() {
 if [ "$1" = "purge" ]; then
     mounts=$(systemctl list-unit-files | grep '^snap[-.].*\.mount' | cut -f1 -d ' ')
     services=$(systemctl list-unit-files | grep '^snap[-.].*\.service' | cut -f1 -d ' ')
-    for unit in $services $mounts; do
-        # ensure its really a snapp mount unit or systemd unit
+
+    for unit in $mounts; do
+        # ensure its really a snap mount unit or systemd unit
```
Comment (niemeyer) [inline debian/snapd.postrm]: Why is this changing? It's unmounting the underlying filesystem from the services before it stops the services? The original logic handled both in the same loop, accounting for what each one is. Not clear why we need to split it up.

File: spread.yaml (hunk 19 above)
Comment (niemeyer) [inline spread.yaml]: Nit: can you please restore the original form?

File: interfaces/builtin/all.go (hunk 24 above)
Comment (vosst) [inline interfaces/builtin/all.go]: I'm fine with undoing this change. One possible approach would be to fail connection of the interface and removing it from the list of implicitly defined slots.

File: debian/control [hunk 26]
```diff
@@ -40,10 +40,13 @@ Package: snapd
 Architecture: any
 Depends: adduser,
          ca-certificates,
+         cgroup-lite,
```
Comment (mvo5) [inline debian/control]: This is a universe package in xenial. So we will need to either get a MIR or only add this dependency on trusty (which will be slightly ugly using debian/control.in (but maybe nowdays there is a better way?).

File: snap/implicit.go (hunk 23 above)
Comment (vosst) [inline snap/implicit.go]: Done, thanks for the suggestion.

File: debian/control [hunk 27]
```diff
@@ -54,11 +54,14 @@ Architecture: any
 Depends: adduser,
          apparmor (>= 2.10.95-0ubuntu2.2),
          ca-certificates,
+         cgroup-lite,
```
Comment (mvo5) [inline debian/control]: This is a universe package in xenial, we either need a SRU or we need to add it only on trusty. Plus it pulls in mountall which is universe as well and not something we want in the default install AIUI (needs double checking with foundations).

File: debian/control [hunk 28]
```diff
@@ -54,11 +54,14 @@ Architecture: any
...
          gnupg1 | gnupg,
          snap-confine (= ${binary:Version}),
          squashfs-tools,
-         systemd,
+         sudo (>=1.8.9p5-1ubuntu1.3),
+         systemd(>=204-5ubuntu20.20),
          ubuntu-core-launcher (= ${binary:Version}),
+         util-linux (>=2.20.1-5.1ubuntu20.9),
```
Comment (mvo5) [inline debian/control]: This package is `Essential: yes` so we don't need to depend on it. We only need it to ensure the version number on trusty is right. Would be nice to add a small comment in here `# needed to pull in the right version on trusty` or something.

File: debian/control [hunk 29]
```diff
@@ -54,11 +54,14 @@ Architecture: any
 Depends: adduser,
          apparmor (>= 2.10.95-0ubuntu2.2),
          ca-certificates,
+         cgroup-lite,
          gnupg1 | gnupg,
          snap-confine (= ${binary:Version}),
          squashfs-tools,
-         systemd,
+         sudo (>=1.8.9p5-1ubuntu1.3),
```
Comment (mvo5) [inline debian/control]: This makes me slightly uneasy. We don't really need sudo for snapd feels not right to list it here just to get the right version pulled in. I guess we would need a conditional dependency here: if-installed: sudo (>= ...) or something. Not sure what we can do to fix this. One way would be a control.in that generates different control files depending on the release in the clean target. This would also solve the util-linux nitpick below.

File: debian/snap.mount.service (hunk 11 above)
Comment (mvo5) [inline debian/snap.mount.service]: Will we have any side-effects here if running snapd inside lxd ?

File: debian/snapd.postinst [hunk 30]
```diff
@@ -11,4 +11,10 @@ case "$1" in
         if dpkg --compare-versions "$2" lt-nl "2.0.7"; then
             ldconfig
         fi
+        # start required services
+        if [ -d /run/systemd/system ]; then
+            systemctl daemon-reload
+            systemctl -f enable snapd.refresh.timer snapd.refresh.service snapd.autoimport.service snapd.socket snapd.service snap.mount.service
```
Comment (mvo5) [inline debian/snapd.postinst]: The debhelper generated scripts use deb-systemd-helper which is doing more than this, it also keeps some persistent state. It makes me slightly uneasy that we are not using this.

File: debian/snapd.postrm (hunk 18 above)
Comment (mvo5) [inline debian/snapd.postrm]: According to the man-page util-linux will be doing the right thing in newer version so a comment that we only have this for trusty compat in here would be nice. We may still need to add the `-d` in `removeMountUnit()` when we remove the snap. If we need the -d here we probably need it also in the go code.

File: debian/snapd.prerm [hunk 31]
```diff
@@ -0,0 +1,8 @@
+#! /bin/sh
+
+set -e
+
+if [ -d /run/systemd/system ]; then
+    systemctl stop snapd.refresh.timer snapd.refresh.service snapd.autoimport.service snapd.socket snapd.service
```
Comment (mvo5) [inline debian/snapd.prerm]: Same comment as in the postinst about the usage of deb-systemd-helper.

File: snap/implicit.go [hunk 32]
```diff
@@ -102,6 +101,9 @@ func AddImplicitSlots(snapInfo *Info) {
 			snapInfo.Slots[ifaceName] = makeImplicitSlot(snapInfo, ifaceName)
 		}
 	}
+	if !(release.ReleaseInfo.ID == "ubuntu" && release.ReleaseInfo.VersionID == "14.04") {
```
Comment (mvo5) [inline snap/implicit.go]: A small comment here why fuse-support is disabled on 14.04 woudl be nice.

File: spread.yaml [hunk 33]
```diff
@@ -149,18 +159,31 @@ prepare: |
...
     trap "sysctl -w net.ipv6.conf.all.disable_ipv6=0" EXIT
 
+    apt-get update
+    apt-get install -y software-properties-common
+
+    if [ "$SPREAD_SYSTEM" = "ubuntu-14.04-64" ]; then
+         echo 'deb http://archive.ubuntu.com/ubuntu/ trusty-proposed main universe' >> /etc/apt/sources.list
+         apt-get update
+
+         add-apt-repository ppa:thomas-voss/trusty
```
Comment (mvo5) [inline spread.yaml]: What is in the PPA that is not in trusty-proposed yet?

File: debian/snap.mount.service (hunk 11 above)
Comment (vosst) [inline debian/snap.mount.service]: I haven't tested that, on the list now.

File: debian/snapd.postrm (hunk 18 above)
Comment (vosst) [inline debian/snapd.postrm]: Done, thanks for the suggestion.

File: debian/control (hunk 29 above)
Comment (vosst) [inline debian/control]: Agreed, we should investigate into a cleaner solution. I would rather like to postpone that and left a comment instead. I hope that's okay with you.

File: debian/control (hunk 28 above)
Comment (vosst) [inline debian/control]: Added comment, see my reply to your previous comment.

File: debian/snapd.postinst (hunk 30 above)
Comment (vosst) [inline debian/snapd.postinst]: Sure, we are running into an issue with the trusty version of systemd being significantly different to the xenial one, though. This is assuming a stable interface in terms of cli, not in terms of side-effects like install location of files.

File: tests/main/chattr/task.yaml [hunk 34]
```diff
@@ -2,6 +2,7 @@ summary: test chattr
 # ubuntu-core doesn't have go :-)
 systems: [-ubuntu-core-16-64, -ubuntu-core-16-arm-64, -ubuntu-core-16-arm-32]
 prepare: |
+  export PATH=/usr/lib/go-1.6/bin:${PATH}
```
Comment (vosst) [inline tests/main/chattr/task.yaml]: ack, I removed the per-test exports accordintly.

File: tests/main/gccgo/task.yaml (hunk 21 above)
Comment (vosst) [inline tests/main/gccgo/task.yaml]: No, we did not, yet.

File: snap/export_test.go [hunk 35]
```diff
@@ -19,12 +19,27 @@
...
-	NewHookType                  = newHookType
+	NewHookType = newHookType
 )
 
+func ImplicitSlotsForTests() []string {
+	result := implicitSlots
+	// fuse-support is disabled on trusty due to usage of fuse requiring access to mount.
+	// we do not want to widen the apparmor profile defined in fuse-support to support trusty
+	// right now.
+	if !(release.ReleaseInfo.ID == "ubuntu" && release.ReleaseInfo.VersionID == "14.04") {
```
Comment (niemeyer) [inline snap/export_test.go]: I don't think we need to do that here, otherwise we're just testing the test logic itself. We can probably revert the changes to this file completely.

**Comment by mvo5** [comment]
All of this has landed in the ubuntu/14.04 branch and is used to build packages.
